### PR TITLE
Visibility

### PR DESCRIPTION
**Pull request by Macroz**
Support for option list of #730.

Includes changes in application, editor and preview.
Screenshot_2020-01-27_10-02-59
Screenshot_2020-01-27_10-02-28

# Definition of Done / Review checklist

## Reviewability
- [x] link to issue
- ~~[ ] note if PR is on top of other PR~~
- ~~[ ] note if related change in rems-deploy repo~~
- [x] consider adding screenshots for ease of review

## API
- [x] API is documented and shows up in Swagger UI
- [x] API is backwards compatible or completely new
- [x] Events are backwards compatible

## Documentation
- [x] update changelog if necessary
- [ ] add or update docstrings for namespaces and functions
- [ ] components are added to guide page
- [ ] documentation _at least_ for config options (i.e. docs folder)
- ~~[ ] ADR for major architectural decisions or experiments~~

## Different installations
- ~~[ ] new configuration options added to rems-deploy repository~~
- [ ] instance specific translations (i.e. LBR kielivara)

## Testing
- [x] complex logic is unit tested
- [x] valuable features are integration / browser / acceptance tested automatically

## Accessibility
- [ ] all icons have the aria-label attribute
- [ ] all fields have a label
- [ ] errors are linked to fields with aria-describedby
- [ ] contrast is checked
- [ ] conscious decision about where to move focus after an action

## Follow-up
- [x] new tasks are created for pending or remaining tasks
- [x] no critical TODOs left to implement

### CODE REVIEW

File: src/clj/rems/db/form.clj [hunk 1]
```diff
@@ -83,13 +83,23 @@
                                       {:field/id "xyz"}
                                       {:field/id "abc"}])))))
 
+(defn- normalize-field-values [field]
```
Comment (opqdonut) [inline src/clj/rems/db/form.clj]: the name sounds like it's talking about `:field/values`. how about something like `normalize-field-definitions`? or perhaps just `normalize`?

File: src/cljs/rems/administration/create_form.cljs [hunk 2]
```diff
@@ -193,15 +206,42 @@
...
+      (if-not  (supports-options? referred-field)
+        {:field/visibility {:visibility/field :t.form.validation/invalid-value}}
+        (if-not (-> visibility :visibility/field :field/id)
+          {:field/visibility {:visibility/field :t.form.validation/invalid-value}}
+          (if-not (empty? (:visibility/values visibility))
+            (when-not (some (field-option-keys referred-field) (:visibility/values visibility))
+              {:field/visibility {:visibility/values :t.form.validation/invalid-value}})
+            {:field/visibility {:visibility/values :t.form.validation/required}})))
+      {:field/visibility {:visibility/field :t.form.validation/required}})
+    {:field/visibility {:visibility/field :t.form.validation/required}}))
```
Comment (opqdonut) [inline src/cljs/rems/administration/create_form.cljs]: it should be possible to express this as a `cond`

File: src/cljs/rems/administration/create_form.cljs [hunk 3]
```diff
@@ -193,15 +206,42 @@
...
+
+(defn- validate-visibility [field fields]
+  (when-let [visibility (:field/visibility field)]
+    (case (:visibility/type visibility)
+      :always nil
+      :only-if (validate-only-if-field field visibility fields)
+      nil {:field/visibility {:visibility/type :t.form.validation/required}}
+      {:field/visibility {:visibility/type :t.form.validation/invalid-value}})))
+
+(defn- validate-field [field index languages fields]
```
Comment (opqdonut) [inline src/cljs/rems/administration/create_form.cljs]: if this was `validate-fields`, wouldn't need to pass `field`, `index` and `fields` separately

File: test/clj/rems/application/test_commands.clj [hunk 4]
```diff
@@ -534,11 +544,29 @@
              (ok-command application submit-command injections))))
 
     (testing "cannot submit when required fields are empty"
```
Comment (opqdonut) [inline test/clj/rems/application/test_commands.clj]: this is now misleading, perhaps just `(testing "required fields"`?

File: test/cljs/rems/administration/test_create_form.cljs [hunk 5]
```diff
@@ -281,7 +281,60 @@
...
+                     (assoc-in [:form/fields 0 :field/id] 0)
+                     (assoc-in [:form/fields 0 :field/type] :option)
+                     (assoc-in [:form/fields 0 :field/options] [{:key "yes"
+                                                                 :label {:en "en yes"
+                                                                         :fi "fi yes"}}
+                                                                {:key "no"
+                                                                 :label {:en "en no"
+                                                                         :fi "fi no"}}])
+                     (assoc-in [:form/fields 1] {:field/id 1
+                                                 :field/stable-id "stable9999"
```
Comment (opqdonut) [inline test/cljs/rems/administration/test_create_form.cljs]: I think stable-id is no more?

File: test/cljs/rems/administration/test_create_form.cljs [hunk 6]
```diff
@@ -398,7 +451,66 @@
...
```
Comment (opqdonut) [inline test/cljs/rems/administration/test_create_form.cljs]: ditto

File: src/cljs/rems/administration/create_form.cljs (hunk 3 above)
Comment (Macroz) [reply src/cljs/rems/administration/create_form.cljs]: Sounds like a good refactoring to do.

File: src/cljs/rems/administration/create_form.cljs (hunk 2 above)
Comment (opqdonut) [reply src/cljs/rems/administration/create_form.cljs]: did it myself

File: src/cljs/rems/administration/create_form.cljs (hunk 3 above)
Comment (Macroz) [reply src/cljs/rems/administration/create_form.cljs]: Done.

**Pull request merged**